Add authorization features and a Pundit mixin

pedantic-git opened the pull request:
This PR adds basic *authorization* features to Administrate, allowing different admins access to different data/actions.

Users are free to roll their own authorization, or use any authz gems they like, but since I personally think [Pundit](https://github.com/elabs/pundit) is the best of the bunch, I've implemented a controller concern that allows users to get Pundit support in one line.

Pundit isn't added as a dependency, except for the tests.

I've added an authorization.md file to the `docs/` with full instructions... let me know what you think and if you'd like any changes!

pablobm commented on app/controllers/administrate/application_controller.rb:
```diff
@@ -104,6 +107,8 @@ def dashboard
       @_requested_resource ||= find_resource(params[:id])
+      authorize_resource(@_requested_resource)
+      @_requested_resource
     end
```
This breaks the existing memoization. Here's an alternative, saving us from (potentially) checking the authorization multiple times:

```ruby
def requested_resource
  @_requested_resource ||= find_resource(params[:id]).tap do |r|
    authorize_resource(r)
  end
end
```

But then I think... we could go further. If the spec for `authorize_resource` required it to return the resource on success, we could do this:

```ruby
def requested_resource
  @_requested_resource ||= authorize_resource(find_resource(params[:id]))
end
```

In fact that's what Pundit's own `authorize` does, which in turn means that the implementation provided by `Punditize` does this, although not the default, blank implementation.

But I'm probably complicating this too much for too little gain, so I'll be happy with something along the lines of my first alternative.

pablobm commented on spec/controllers/admin/orders_controller_spec.rb:
```diff
@@ -0,0 +1,70 @@
+        locals = capture_view_locals { get :index }
+        expect(locals[:resources].count).to eq(9) # only my orders
+      end
+    end
```
Please separate these with blank lines. I'm surprised that Hound is not complaining about this and the single quotes?

nickcharlton commented:
Thanks @pedantic-git! I know @pablobm left some comments, but I'm thinking it'd be good to get this into `0.9`, which should be in a few weeks time (i.e.: the next release).

pedantic-git commented:
Sure! I'll try to get onto these tomorrow. Just working on *another* Administrate PR right now (trying to get editing working for Field::Polymorphic... nearly there too!)

pedantic-git commented on spec/example_app/app/policies/application_policy.rb:
```diff
@@ -0,0 +1,53 @@
+  end
+
+  def show?
+    scope.where(:id => record.id).exists?
```
I'd rather not address this comment because application_policy.rb is included verbatim as generated by the "pundit:install" generator.

pedantic-git commented on app/controllers/administrate/application_controller.rb:
```diff
@@ -145,5 +148,20 @@ def show_search_bar?
       ).any? { |_name, attribute| attribute.searchable? }
     end
+
+    def show_action?(action, resource)
```
Not addressing this - it's a method signature that is meant to be overridden in subclasses so the arguments and their names are important.

pedantic-git commented:
OK! I've addressed all but two of the Hound violations (see above for my reasoning), I've changed the way `authorize_resource` works (thanks for spotting this, @pablobm!) and I've removed the comments.

[Removing comments is still something I struggle with - I can understand other elements of house style that I disagree with but this one always seems very counterintuitive to me.]

Let me know if I can do anything else!

pedantic-git commented:
Just realized that Pundit 1.1.0 (the latest gem) doesn't return the object from `authorize` - it returns `true`. Shocking there hasn't been a gem release in over a year, but in the meantime I'll change the `requested_resource` to use tap instead.

pustomytnyk commented on app/views/administrate/application/edit.html.erb:
```diff
@@ -27,7 +27,7 @@ It displays a header, and renders the `_form` partial to do the heavy lifting.
       [namespace, page.resource],
       class: "button",
-    ) if valid_action? :show %>
+    ) if valid_action?(:show) && show_action?(:show, page.resource) %>
```
Is it possible to use only `show_action?` in all cases?
Like
```ruby
def show_action?(action, resource)
  valid_action?(action, resource) && authorized?(action, resource)
end
```

pedantic-git replied:
Hi @pustomytnyk - `#show_action?` is a method that users of administrate can (and should) override in their controllers. If we implement it in terms of `#valid_action?` in the base controller and again in the Pundit mixin, there's still opportunity for administrate users to accidentally break the expected behaviour by overriding it to something that's not.

nickcharlton commented:
Thanks, @pedantic-git! I'm (finally!) going to merge this now.

pedantic-git commented:
Yay! Thanks @nickcharlton !!
